src/main/java/org/jenkinsci/plugins/workflow/support/steps/StageStepExecution.java

```diff
@@ -54,12 +60,34 @@
 
     @Override
     public boolean start() throws Exception {
+        if (isInsideParallel(node)) {
+            throw new AbortException(Messages.StageStepExecution_the_stage_step_must_not_be_used_inside_a());
+        }
         node.addAction(new LabelAction(step.name));
         node.addAction(new StageActionImpl(step.name));
         enter(run, getContext(), step.name, step.concurrency);
         return false; // execute asynchronously
     }
 
+    // TODO switch to FlowNodeSerialWalker or equivalent from https://github.com/jenkinsci/workflow-api-plugin/pull/2
+    private static boolean isInsideParallel(FlowNode n) {
+        while (true) {
+            if (n instanceof BlockEndNode) {
+                n = ((BlockEndNode) n).getStartNode();
+            }
+            if (n.getAction(ThreadNameAction.class) != null) {
+                return true;
+            }
+            List<FlowNode> parents = n.getParents();
+            if (parents.isEmpty()) {
+                assert n instanceof FlowStartNode;
+                return false;
+            }
+            assert parents.size() == 1;
```

Comment:
🐜 I'm not *super* thrilled by the assertion use here vs. throwing a more explicit IllegalStateException.